ingest.py

```diff
@@ -79,15 +79,17 @@ def main():
     chunk_overlap = 50
     documents = load_documents(source_directory)
     text_splitter = RecursiveCharacterTextSplitter(chunk_size=chunk_size, chunk_overlap=chunk_overlap)
-    texts = text_splitter.split_documents(documents)
+    text_docs = text_splitter.split_documents(documents)
+    texts = [doc.page_content for doc in text_docs]
+    metadatas = [doc.metadata for doc in text_docs]
```

Comment:
It seems a bit wasteful to loop two times through `text_docs` (NIT: BTW we could just call it `docs`), instead of just once. On the other hand, this code comes from `langchain` directly, so maybe it's worth keeping it as is just for convenience sake.

Reply:
Among the python priesthood, list comprehensions are considered more concise and readable than 'for' loops. I don't see any claim that they are more efficient, though. Were it not for the 'original' mechanism, I would have written a 'for' loop.

On the var name, I am not feeling much joy with calling the chunks produced by the splitter 'docs' or, originally, 'documents' because, despite having the type they are very distinct from the original documents. Text_docs was a compromise of function 'text' and type 'doc'. On further consideration, I would choose chunks for clarity.

While I am whining about things, I don't think we are getting very good service from the loader-splitter combinations in terms of metadata. The splitters take the raw text and apply the same metadata from the document to each chunk. That is fine for a two or ten page document, but it is not very helpful to know that the relevant passage is somewhere in a 500 page book when it comes time to refer to, quote that snippet or create a citation.
The loaders know about pages, but don't add anything to the output to identify where the boundaries are and the splitters don't even add the offset into the text conversion to the metadata.
And yes, I have quite a few 500 page books in my digital collection.
</whining>

Reply:
Oh I prefer list comprehension too, but intuitively it doesn't make sense. Tho, you make a very good point about performance. Because of how Python works, it's quite possibly much faster than even just a single `for` loop .

`for chunk in chunks` works very well for me too.

Reply:
@PulpCattel Thanks for your attentions. I realized that time has marched on and I probably need to merge these changes into the now much different ingest.py on the main branch.
That will give me the chance to introduce new errors in the process but I will try to keep it to a minimum. I don't know if I need to put a pause on this request beyond this notice or do anything else while I work. Not a big deal but a pretty big change for the appearance of the pull request.

Reply:
I explored the more current code and it is not really compatible with this strategy. However, it offers much greater opportunity for dealing with this case and related issues.